pkg/registries/registries.go

```diff
@@ -35,81 +36,208 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
-				topoGraph.AddEdge(set.Mirrors[i], set.Mirrors[i+1])
-			}
-			sourceInGraph := false
+		source, mirrors, err := digestMirrorSets.mergedMirrors(source)
+		if err != nil {
+			return nil, err
+		}
+		imageDigestMirror := apicfgv1.ImageDigestMirrors{
+			Source:  source,
+			Mirrors: mirrors,
+		}
+		if digestMirrorSets.mirrorBlockSource[source] {
+			imageDigestMirror.MirrorSourcePolicy = apicfgv1.NeverContactSource
+		}
+		res = append(res, imageDigestMirror)
+
+	}
+	return res, nil
+}
+
+func mergedICSPMirrorSets(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy) ([]apicfgv1.ImageDigestMirrors, error) {
+	repositoryMirrorSets := newMirrorSets()
+	for _, icsp := range icspRules {
+		for i := range icsp.Spec.RepositoryDigestMirrors {
+			set := icsp.Spec.RepositoryDigestMirrors[i]
+			imgMirrors := []apicfgv1.ImageMirror{}
 			for _, m := range set.Mirrors {
-				if m == source {
-					sourceInGraph = true
-					break
-				}
+				imgMirrors = append(imgMirrors, apicfgv1.ImageMirror(m))
 			}
-			if !sourceInGraph {
-				// The build of mirrorSets guarantees len(set.Mirrors) > 0.
-				topoGraph.AddEdge(set.Mirrors[len(set.Mirrors)-1], source)
-			}
-			// Every node in topoGraph, including source, is implicitly added by topoGraph.AddEdge (every mirror set contains at least one non-source mirror,
-			// so there are no unconnected nodes that we would have to add separately from the edges).
+			repositoryMirrorSets.addMirrorSets(set.Source, "", imgMirrors)
 		}
-		sortedRepos, err := topoGraph.Sorted()
+	}
+
+	// Sort the sets of mirrors by Source to ensure deterministic output
+	sources := repositoryMirrorSets.sourceList()
+	// Convert the sets of mirrors
+	res := []apicfgv1.ImageDigestMirrors{}
+	for _, source := range sources {
+		source, mirrors, err := repositoryMirrorSets.mergedMirrors(source)
 		if err != nil {
 			return nil, err
 		}
-		if sortedRepos[len(sortedRepos)-1] == source {
-			// We don't need to explicitly include source in the list, it will be automatically tried last per the semantics of sysregistriesv2. Mirrors.
-			sortedRepos = sortedRepos[:len(sortedRepos)-1]
-		}
-		res = append(res, apioperatorsv1alpha1.RepositoryDigestMirrors{
+		repositoryMirror := apicfgv1.ImageDigestMirrors{
+			// leave MirrorSourcePolicy black, it will follow the default AllowContactingSource
 			Source:  source,
-			Mirrors: sortedRepos,
-		})
+			Mirrors: mirrors,
+		}
+		res = append(res, repositoryMirror)
+
 	}
 	return res, nil
 }
 
+func topoSortRepos(source string, ds *[][]apicfgv1.ImageMirror) ([]string, error) {
+	topoGraph := newTopoGraph()
+	for _, mirrors := range *ds {
+		for i := 0; i+1 < len(mirrors); i++ {
+			topoGraph.AddEdge(string(mirrors[i]), string(mirrors[i+1]))
+		}
+		sourceInGraph := false
+		for _, m := range mirrors {
+			if string(m) == source {
+				sourceInGraph = true
+				break
+			}
+		}
+		if !sourceInGraph {
+			// The build of mirrorSets guarantees len(set.Mirrors) > 0.
```

Comment:
```suggestion
			// mirrorSets.addMirrorSets guarantees len(set.Mirrors) > 0.
```
or, alternatively, make this, or the loop calling this, a method on `mirrorSets`.